File: sdk/samples/WCFSamples/TechnologySamples/Scenario/Rules/OrderProcessingPolicy/CS/OrderErrorLibrary/OrderErrorLibrary.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.Collections.ObjectModel;

namespace Microsoft.Rules.Samples
{
    // The container class of the extension method.
    public static class Extensions
    {
        // The definition of the extension method for the class OrderErrorCollection.
        public static void PrintOrderErrors(this OrderErrorCollection orderErrorCollection)
        {
            Console.WriteLine();

            foreach (OrderError orderError in orderErrorCollection)
            {
                Console.WriteLine(orderError.ErrorText);
            }
        }
    }

    // The error object that gets created when an invalid input is entered.
    public class OrderError
    {
        private string errorText;       // that is, Error: Zip code is invalid.
        private string customerName;    // that is, "John Customer".
        private int itemNum;            // that is, 1 => for Vista Ultimate DVD.
        private string zipCode;         // that is, "00999".

        public OrderError()
        {
        }

        public OrderError(int invalidItemNum)
        {
            this.ItemNum = invalidItemNum;
        }

        public OrderError(string invalidZipCode)
        {
            this.ZipCode = invalidZipCode;
        }

        public string ErrorText
        {
            get
            {
                return this.errorText;
            }
            set
            {
                this.errorText = value;
            }
        }

        public string CustomerName
        {
            get
            {
                return this.customerName;
            }
            set
            {
                this.customerName = value;
            }
        }

        public int ItemNum
        {
            get
            {
                return this.itemNum;
            }
            set
            {
                this.itemNum = value;
            }
        }

        public string ZipCode
        {
            get
            {
                return this.zipCode;
            }
            set
            {
                this.zipCode = value;
            }
        }
    }

    public class OrderErrorCollection : Collection<OrderError>
    {
        public OrderErrorCollection()
        {
        }

        public void AddError(OrderError orderError)
        {
            this.Add(orderError);
        }

        // Overload the operator + for two OrderErrorCollection objects.
        public static OrderErrorCollection operator +(OrderErrorCollection orderErrorCollection1, OrderErrorCollection orderErrorCollection2)
        {
            OrderErrorCollection orderErrorCollection = new OrderErrorCollection();
            if (null != orderErrorCollection1)
            {
                foreach (OrderError orderError in orderErrorCollection1)
                    orderErrorCollection.Add(orderError);
            }

            if (null != orderErrorCollection2)
            {
                foreach (OrderError orderError in orderErrorCollection2)
                    orderErrorCollection.Add(orderError);
            }

            return orderErrorCollection;
        }
    }
}

File: sdk/samples/WCFSamples/TechnologySamples/Scenario/Rules/OrderProcessingPolicy/VB/OrderErrorLibrary/OrderErrorLibrary.vb
' Copyright (c) Microsoft Corporation. All rights reserved.

Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Text
Imports System.Runtime.CompilerServices

Namespace Microsoft.Rules.Samples
    ' The container class of the extension method
    Public Module Extensions
        ' The definition of the extension method for the class OrderErrorCollection
            <Extension()> _
            Public Sub PrintOrderErrors(ByVal orderErrorCollection As OrderErrorCollection)
            Console.WriteLine()

            For Each orderError As OrderError In orderErrorCollection
                Console.WriteLine(orderError.ErrorText)
            Next
        End Sub
    End Module

' The error object that will get created when an invalid input is entered
Public Class OrderError
    Private m_errorText As String ' ie. Error: Zip code is invalid
    Private m_customerName As String ' ie. "John Customer"
    Private m_itemNum As Integer ' ie. 1 => for Vista Ultimate DVD
    Private m_zipCode As String ' ie. "00999"
    Public Sub New()
    End Sub

    Public Sub New(ByVal invalidItemNum As Integer)
        Me.ItemNum = invalidItemNum
    End Sub

    Public Sub New(ByVal invalidZipCode As String)
        Me.ZipCode = invalidZipCode
    End Sub

    Public Property ErrorText() As String
        Get
            Return Me.m_errorText
        End Get
        Set(ByVal value As String)
            Me.m_errorText = value
        End Set
    End Property

    Public Property CustomerName() As String
        Get
            Return Me.m_customerName
        End Get
        Set(ByVal value As String)
            Me.m_customerName = value
        End Set
    End Property

    Public Property ItemNum() As Integer
        Get
            Return Me.m_itemNum
        End Get
        Set(ByVal value As Integer)
            Me.m_itemNum = value
        End Set
    End Property

    Public Property ZipCode() As String
        Get
            Return Me.m_zipCode
        End Get
        Set(ByVal value As String)
            Me.m_zipCode = value
        End Set
    End Property
End Class

Public Class OrderErrorCollection
    Inherits Collection(Of OrderError)
    Public Sub New()
    End Sub

    Public Sub AddError(ByVal orderError As OrderError)
        Me.Add(orderError)
    End Sub

    ' Overload the operator + for two OrderErrorCollection objects
    Public Shared Operator +(ByVal orderErrorCollection1 As OrderErrorCollection, ByVal orderErrorCollection2 As OrderErrorCollection) As OrderErrorCollection
        Dim orderErrorCollection As New OrderErrorCollection()
        If orderErrorCollection1 IsNot Nothing Then
            For Each orderError As OrderError In orderErrorCollection1
                orderErrorCollection.Add(orderError)
            Next
        End If

        If orderErrorCollection2 IsNot Nothing Then
            For Each orderError As OrderError In orderErrorCollection2
                orderErrorCollection.Add(orderError)
            Next
        End If

        Return orderErrorCollection
    End Operator
End Class
End Namespace
